### PR TITLE
Keyword search

### PR DESCRIPTION
**Pull request by 0tpyrk**
Added a separate search servlet that implements a keyword search/sorting algorithm. Also fixed some random bugs having to do with search.

Still needs to be done:
- Frontend
- Generation of keywords in event creation/editing

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        map.put(w, map.get(w) + 1);
+      } else {
+        map.put(w, 1);
+      }
+    }
+    return map;
+  }
+
+  /**
+   * Returns a the total number of weighted occurrences of keywords in a list.
```
Comment (Yunni) [inline src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:293]: The doc seems very ambiguous. Please give a formula or example. The param names could be improved as well.

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+   * Returns the keywords for a map.
+   *
+   * @param map Map containing the keywords and their weighted occurrences
+   * @return List containing the keywords only
+   */
+  public static List<String> getKeywordMapKeys(Map<String, Integer> map) {
+    List<String> list = new ArrayList<String>();
+    List<Map.Entry<String, Integer>> mapList =
+        new ArrayList<Map.Entry<String, Integer>>(map.entrySet());
+    Collections.sort(
```
[the remark this answers is not in the document]
Comment (0tpyrk) [reply src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:212]: The sort is unique to the Keyword search (the only other place it's currently used is in the tags search, but even that is modified significantly) so should it still be moved?

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java (hunk 1 above)
Comment (0tpyrk) [reply src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:293]: Got it!

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+   * @return Map containing word keys and instance count values
+   */
+  public static Map<String, Integer> wordCount(String input) {
+    List<String> words = KeywordSearchServlet.getSeparateWords(input);
+    // process title words
+    Map<String, Integer> map = new HashMap<String, Integer>();
+    for (String w : words) {
+      w = w.toLowerCase();
+      if (map.containsKey(w)) {
+        map.put(w, map.get(w) + 1);
```
[the remark this answers is not in the document]
Comment (0tpyrk) [reply src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:284]: Got it!

File: src/main/java/com/google/sps/servlets/SearchServlet.java [hunk 4]
```diff
@@ -118,16 +108,17 @@ public int compare(Entity o1, Entity o2) {
...
             int compareSize = Integer.compare(o2List.size(), o1List.size());
             if (compareSize != 0) {
               return compareSize;
-            } else {
-              // Sort by which event is closer to the user
-              return Integer.compare(
-                  Integer.parseInt(o1.getProperty("distance").toString()),
-                  Integer.parseInt(o2.getProperty("distance").toString()));
-            }
+            } else {*/
```
[the remark this answers is not in the document]
Comment (0tpyrk) [reply src/main/java/com/google/sps/servlets/SearchServlet.java:116]: Oops, thanks!

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+            Math.round(
+                Integer.parseInt(Utils.getParameter(request, "searchDistance", "")) * MI_TO_KM));
+    // Remove from the list any events outside the cutoff or that aren't drivable
+    events.removeIf(
+        e -> ((int) e.getProperty("distance")) > cutoff || ((int) e.getProperty("distance")) < 0);
+
+    // Sort list by most keywords in common with search
+    Collections.sort(
+        events,
+        new Comparator<Entity>() {
```
[the remark this answers is not in the document]
Comment (0tpyrk) [reply src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:108]: Got it!

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yunni) [reply src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:218]: It looks very similar to L180 and L241. Please at least parameterize it.

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+  private static final List<String> IRRELEVANT_WORDS =
+      new ArrayList<String>(Arrays.asList("the", "is", "for", "in", "of", "so", "to"));
+  // Conversion rate from mi to km
+  private static final double MI_TO_KM = 1.609;
+  // Keywords retrieved from the frontend in the doGet method
+  private List<String> searchKeywords;
+  /**
+   * Comparator used to compare the relevance of two events to the user and their search keywords.
+   */
+  private final Comparator<Entity> KEYWORD_SEARCH_RELEVANCE =
```
Comment (Yunni) [inline src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:61]: Please fix.

File: src/main/java/com/google/sps/servlets/KeywordSearchServlet.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (0tpyrk) [reply src/main/java/com/google/sps/servlets/KeywordSearchServlet.java:218]: Oops, I was looking at the wrong sort. Yes, I'm on it!

**Pull request merged**